forgerock-authenticator/src/main/java/org/forgerock/android/authenticator/Mechanism.java

```diff
@@ -95,6 +98,34 @@ public String getSecret() {
         return secret;
     }
 
+    @Override
+    public abstract String toJson();
+
+    /**
+     * Deserializes the specified Json into an object of the {@link Mechanism} object.
+     * @param jsonString the json string representing the object to be deserialized
+     * @return an {@link Mechanism} object from the string. Returns {@code null} if {@code jsonString} is {@code null}
+     * or if {@code jsonString} is empty.
+     */
+    public static Mechanism fromJson(String jsonString) {
+        Mechanism mechanism = null;
+        if (jsonString == null || jsonString.length() == 0) {
+            return null;
+        }
+        try {
+            JSONObject jsonObject = new JSONObject(jsonString);
+            String type = jsonObject.getString("type");
+            if(type.equals(PUSH)) {
+                mechanism = Push.fromJson(jsonString);
+            } else if(type.equals(OATH)) {
+                mechanism = Oath.fromJson(jsonString);
+            }
```

Comment:
What if not PUSH and OATH? should we throw exception? or expect null?

Reply:
It can only be PUSH or OATH at this time. It is expected to be null if not able to parse.